quickjs.c

```diff
@@ -45531,180 +45531,180 @@ static int js_json_to_str(JSContext *ctx, JSONStringifyContext *jsc,
-                    }
-                    string_buffer_concat_value(jsc->b, sep);
-                    string_buffer_concat_value(jsc->b, prop);
-                    string_buffer_putc8(jsc->b, ':');
-                    string_buffer_concat_value(jsc->b, sep1);
-                    if (js_json_to_str(ctx, jsc, val, v, indent1))
-                        goto exception;
-                    has_content = true;
                 }
+
+                JS_FreeValue(ctx, prop_str);
+
+                if (js_json_to_str(ctx, jsc, val, prop_val, indent1) < 0)
+                    goto exception;
             }
-            if (has_content && JS_VALUE_GET_STRING(jsc->gap)->len != 0) {
-                string_buffer_putc8(jsc->b, '\n');
-                string_buffer_concat_value(jsc->b, indent);
-            }
-            string_buffer_putc8(jsc->b, '}');
+
+            if (string_buffer_putc8(jsc->b, '}') < 0)
+                goto exception;
+
+            js_free_prop_enum(ctx, props, prop_count);
         }
-        if (check_exception_free(ctx, js_array_pop(ctx, jsc->stack, 0, NULL, 0)))
-            goto exception;
-        JS_FreeValue(ctx, val);
-        JS_FreeValue(ctx, tab);
+
+        JS_FreeValue(ctx, indent1);
         JS_FreeValue(ctx, sep);
         JS_FreeValue(ctx, sep1);
-        JS_FreeValue(ctx, indent1);
+        JS_FreeValue(ctx, tab);
         JS_FreeValue(ctx, prop);
         return 0;
+
+exception:
+        JS_FreeValue(ctx, indent1);
+        JS_FreeValue(ctx, sep);
+        JS_FreeValue(ctx, sep1);
+        JS_FreeValue(ctx, tab);
+        JS_FreeValue(ctx, prop);
+        return -1;
     }
- concat_primitive:
+
     switch (JS_VALUE_GET_NORM_TAG(val)) {
-    case JS_TAG_STRING:
-        val = JS_ToQuotedStringFree(ctx, val);
-        if (JS_IsException(val))
-            goto exception;
-        goto concat_value;
-    case JS_TAG_FLOAT64:
-        if (!isfinite(JS_VALUE_GET_FLOAT64(val))) {
-            val = JS_NULL;
+        case JS_TAG_STRING: {
+            JSValue quoted = JS_ToQuotedStringFree(ctx, js_dup(val));
+            if (JS_IsException(quoted))
+                return -1;
+            return string_buffer_concat_value(jsc->b, quoted);
         }
-        goto concat_value;
-    case JS_TAG_INT:
-    case JS_TAG_BOOL:
-    case JS_TAG_NULL:
-    concat_value:
-        return string_buffer_concat_value_free(jsc->b, val);
-    case JS_TAG_BIG_INT:
-        JS_ThrowTypeError(ctx, "BigInt are forbidden in JSON.stringify");
-        goto exception;
-    default:
-        JS_FreeValue(ctx, val);
-        return 0;
+        
+        case JS_TAG_FLOAT64: {
+            double d = JS_VALUE_GET_FLOAT64(val);
+            if (!isfinite(d)) {
+                return string_buffer_concat_value(jsc->b, JS_NULL);
+            }
+            JSValue num_str = JS_ToString(ctx, val);
+            if (JS_IsException(num_str))
+                return -1;
+            return string_buffer_concat_value(jsc->b, num_str);
```

Comment:
This is a **memory leak**: should be `return string_buffer_concat_value_free(jsc->b, num_str)`